[RFC FS-1151] Record spreads

brianrourkeboll opened the pull request:
RFC for the record-to-record spreads subset of the approved language suggestion https://github.com/fsharp/fslang-suggestions/issues/1253.

- Discussion thread (covers more than just records): https://github.com/fsharp/fslang-design/discussions/806
- Implementation: https://github.com/dotnet/fsharp/pull/18927

Click “Files changed” → “⋯” → “View file” for the rendered RFC.

brianrourkeboll committed:
RFC FS-1151 - Record spreads

T-Gro commented on RFCs/FS-1151-spread-operator.md:
```diff
@@ -0,0 +1,14 @@
+
+- [x] [Spread operator for F#](https://github.com/fsharp/fslang-suggestions/issues/1253)
+- [x] Approved in principle
+- [ ] [Implementation](https://github.com/dotnet/fsharp/pull/TODO)
```
Point this at the implementation PR:

```suggestion
- [ ] [Implementation](https://github.com/dotnet/fsharp/pull/18927)
```

T-Gro commented on RFCs/FS-1151-spread-operator.md:
```diff
@@ -0,0 +1,14 @@
+
+# Summary
+
+We add support in the following specific contexts for spread expressions of the form `...source`, wherein the symbol `...` prefixed to a `source` expression or type indicates that a set of values or members is to be copied or "spread" from the value or type represented by the source expression into the target context in some way.
```
This needs a body before it can leave draft. Minimal spine below: Summary/scope, Motivation, Detailed design with the shadowing example, the grammar delta for the spec, and Compatibility. I tightened your Summary line in the process. Deliberately skipped Drawbacks/Alternatives (one-liners at most — `..` vs `...` is settled in #806) and the full diagnostics list (it's in #18927 if you'd rather enumerate it).

```suggestion
A spread `...source` copies the fields of a record value or type into another record. This RFC adds spreads, under `--langversion:preview`, in record type definitions — `type T = { ...U; ... }` — and in nominal and anonymous record expressions — `{ ...src; ... }` and `{| ...src; ... |}`. Spreading non-record types, mixing `...` with `with`, and spreads in collection, computation, and pattern positions are out of scope.

# Motivation

`with` copy-and-update cannot change a field's type or combine more than one source, and record types must repeat shared fields explicitly. Spreads remove that duplication.

# Detailed design

`...source` copies the fields of `source` into the surrounding record: inlined field declarations in a type definition, and field assignments in an expression, where the source is evaluated once. Fields combine left to right and a later field shadows an earlier field of the same name; a spread field shadowing an explicit field warns, and two explicit fields of the same name are an error.

    type R1 = { A: int; B: int }
    type R2 = { A: string; ...R1 }  // { A: int; B: int } — spread shadows explicit A (warns)
    type R3 = { ...R1; A: string }  // { A: string; B: int } — explicit shadows spread A

# Changes to the F# spec

A record field declaration and a record/anonymous-record expression binding each gain a spread form:

    recd-field-decl  +=  '...' type      // in record type definitions
    recd-binding     +=  '...' expr      // in record and anonymous record expressions

# Compatibility

Not a breaking change: the syntax is new and rejected by earlier compilers. Emitted assemblies contain ordinary record definitions and construction; no FSharp.Core or runtime support is required.
```

brianrourkeboll committed:
Clarify record-to-record

brianrourkeboll committed:
Add info warnings

brianrourkeboll committed:
Actually, I disagree with these info warnings

We can always add them back later, and if we do, people can treat them
as errors
